### ISSUE TITLE
New syntax for the GRIN IR

### ISSUE THREAD

**Issue by Anabra** [grin-compiler/grin#32]
New syntax proposal
===================

This document proposes new syntactic construct for the GRIN IR, as well as modifications to some of the already existing ones.

Motiviation
-----------

The current syntax of GRIN can pose some difficulties for analyses. As an example the created-by analysis requires each binding to have a variable pattern. That is, each binding must a have _name_. Also, analyzing programs where every intermediate value has an explicit name is much easier.

Furthermore, the current Haskell definition of the syntax allows certain erroneous programs. Currently, we rely on the linter to reject these programs, but a more rigid definition of the syntax could prevent errors statically.

Naming
------

These syntactic constraints mainly deal with naming of intermediate values. They make sure, that most of the currently available values can always be refered to by an explicit name. All of these constraints can be implemented trivially.

  - named case scrutinees
  - named node fields
  - name function arguments
  - named `store` argument

A notable exception is the pattern matching construct for bindings. We will deal with them seperately. Also, note that `fetch` and `update` already only accept variable names as arguments.

New
---

These syntactic constructs are completely new to the GRIN language.

### @patterns

As mentioned earlier, each binding should have a name, the variable it binds. Currently, the syntax allows for pattern bindings, which do not bind a variable to the left-hand side computation. This problem could be fixed by introducing @patterns.

```haskell
v@(CInt k) <- pure n
<both v and k can be referenced here>
```

@patterns combine the flexibility of value patterns with the rigidness of the naming convention. By removing value patterns from the language, and introducing @patterns, we could achieve the same expressive power as before, meanwhile making sure that everything can be referenced by an explicit name.

### Function application

Currently, normal functions and externals share the same syntactic node for application. Externals are stored together with GRIN programs, and they are differentiated from normal functions by looking up the applied function's name in the stored external list. This makes analyzing function applications quite inconvient.

We could introduce different syntactic nodes for normal functions and externals, but that would pose an unnecessary overhead on analyses and transformations in certain use cases. Instead, we will introduce different _types of names_ for functions and externals. The application node will differentiate functions from externals by wrapping teir names in different data constructors.

```
data FName
  = F { fName :: Name }
  | E { fname :: Name }

Exp = ... | SAApp FName [Name] | ...
```

Structural
----------

These modifications impose certain structural constraints on GRIN porgrams.

### Last return

Binding sequences should always end in a `return` (or `pure`). This will make the control flow a little bit more explicit. However, this change could be impacted by the introduction of basic blocks. It might be wiser to delay this change.

### Program, function, definition

The above mentioned notions should be different syntactic constructs. They should form a hierarchy, so it is always clear whether a transformantion/analysis works on an entire GRIN program, a single function, or only on an expression.

Currently available, but has to be more precise
-----------------------------------------------

These modifications would turn currently linter-checked properties into static constraints. For example, only _real_ expressions could appear on the left-hand side of a binding, or the alternatives of a case expression could really only be alternatives.

No longer needed
----------------

By introducing the above mentioned syntactic modifiaction, some currently available constraints become redundant: `LPat`, `SimpleVal`.

Future
------

As for the future, we plan to introduce basic blocks in to the language. This will bring its own syntactic modifications, but they will be mostly independent of the above discussed changes.

**Comment by andorp** [comment]
{code}
named case scrutinees
named node fields
name function arguments
named store argument
{code}
Please could you give examples how they are now and what will be the intended result?
We have named function arguments, probably you meant at the call side.

**Comment by andorp** [comment]
```
data FName
  = F { fName :: Name }
  | E { fname :: Name }
```
How FName will fit to the current analyses? Currently we have unified Names everywhere. When projecting back to source code, type env, should this introduce difficulties?

**Comment by andorp** [comment]
Last return

Lets use the `pure` as it is already established in the syntax, as also doesn't give the
false sense of return point of a function.

**Comment by andorp** [comment]
```
Program, function, definition
The above mentioned notions should be different syntactic constructs.
```

This design decision was taken to be compatible with the recursion schemes library, which already
had the tradeoff. Combining base functors would leave the intermix of the different levels.
Using GADTs and indexed recursion schemes would mean reimplementing the recursion-schemes library, which support indexes.

I was considering introduce a type safe layer which would use a GADT to construct only meaningful programs, but I rejected the idea out of pure lazyness (pun intended). Maybe it is a good idea to have that layer for the future frontends, forcing them (another pun intended) to create a correct syntax. But for us GRIN contributors I think it is overkill to create such a type-safety when implementing transformations, analyses, at least for now, as we plan to refactor the IR. If we want to do such a type safe internal representation, I would suggest to implement it when we plan to do the first official release.

**Comment by Anabra** [comment]
> ```
> named case scrutinees
> named node fields
> name function arguments
> named store argument
> ```
> 
> Please could you give examples how they are now and what will be the intended result?
> We have named function arguments, probably you meant at the call side.

the ar

**Comment by Anabra** [comment]
> ```
> data FName
>   = F { fName :: Name }
>   | E { fname :: Name }
> ```
> 
> How FName will fit to the current analyses? Currently we have unified Names everywhere. When projecting back to source code, type env, chould this introduce difficulties?

`FName` would only be present in the application node, nowhere else. This way, if the analysis/transformation does not want to differentiate normal functio apps from external apps, they could just pattern patch on `App` and get the stored name via the getter. If someone do want to differentiate them, they will be able to pattern patch on the respective `FName` data ctors.

During pretty printing, I don't think it would introduce any difficulties. These are just wrappers.

**Comment by Anabra** [comment]
> Last return
> 
> Lets use the `pure` as it is already established in the syntax, as also doesn't give the
> false sense of return point of a function.

Alright.

**Comment by Anabra** [comment]
> ```
> Program, function, definition
> The above mentioned notions should be different syntactic constructs.
> ```
> 
> This design decision was taken to be compatible with the recursion schemes library, which already
> had the tradeoff. Combining base functors would leave the intermix of the different levels.
> Using GADTs and indexed recursion schemes would mean reimplementing the recursion-schemes library supporting indexes.
> 
> I was considering introduce a type safe layer which would use a GADT to construct only meaningful programs, but I rejected the idea out of pure laziness (pun intended). Maybe it is a good idea to have that layer for the future frontends, forcing them (another pun intended) to create a correct syntax. But for us GRIN contributors I think it is overkill to create such a type-safety when implementing transformations, analyses, at least for now, as we plan to refactor the IR. If we want to do such a type safe internal representation, I would suggest to implement it when we plan to do the first official release.

Makes sense. I did not consider this.

**Comment by andorp** [comment]
Naming:
I see, basically you suggest to introduce constants using the `pure` construction. `v <- pure 1` .

**Comment by andorp** [comment]
Please could you update the description of the proposal?

**Comment by Anabra** [comment]
Exactly. Constants could only be introduced in the above form. Everywhere else, we would be just passing variables around.

**Comment by Anabra** [comment]
The __separation of__ external and ordinary __function application has been scrapped__. The reason being is that it makes writing code for function application cumbersome. Whenever we need the fact that the function is an external, we will usually have to look it up from the external mapping anyway. This lookup will determine whether the function is an external or not. This means, that the pattern match on the type of the function application is redundant.

**Comment by Anabra** [comment]
Extended the proposal with the __removal of low-level GRIN construct__.

`VarTag` node and indexed `Fetch` will be removed. These constructs were originally needed for RISC code generation. Since then, GRIN transitioned to LLVM, and over the years these constructs proved to be unnecessary.

**Comment by andorp** [comment]
As part of the ICFP tutorial there is a GADT, which helps to create syntactically expected GRIN programs. The rule indexed expressions show what can be combined how.

This approach will live in the Syntax.GADT module. This change can be found in the https://github.com/grin-compiler/grin/pull/44 PR.

**Comment by Anabra** [comment]
I added a brief description for the GADT-style AST, and the new datalog based model for the AST. The DL model will need named case alternatives.

**Comment by Anabra** [comment]
Added a detailed explanation for named case alternatives.

**Comment by andorp** [comment]
Looking at the code from the latest PR, we realized that it is confusing to have two different forms for the @as pattern

```
case v of
  (CNil)       @ v1 -> <code>
  (CCons x xs) @ v2 -> <code>

v@(CInt k) <- pure nis 
<both v and k can be referenced here>
```
This should be:

```
case v of
  (CNil)       @ v1 -> <code>
  (CCons x xs) @ v2 -> <code>

(CInt k)@v <- pure nis 
<both v and k can be referenced here>
```

**Comment by andorp** [comment]
In the long run, we should remove the double variable @as pattern construction.

```
var1@var2 <- lhs
```

This should be only
```
var1 <- lhs
```

**Comment by Anabra** [comment]
I wrote a few lines about this in the as-pattern section.
